File: app/src/main/java/piuk/blockchain/android/ui/swipetoreceive/SwipeToReceiveFragment.java
package piuk.blockchain.android.ui.swipetoreceive;


import android.content.BroadcastReceiver;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.databinding.DataBindingUtil;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import piuk.blockchain.android.R;
import piuk.blockchain.android.data.websocket.WebSocketService;
import piuk.blockchain.android.databinding.FragmentSwipeToReceiveBinding;
import piuk.blockchain.android.ui.balance.BalanceFragment;
import piuk.blockchain.android.ui.customviews.ToastCustom;
import piuk.blockchain.android.util.OSUtil;
import piuk.blockchain.android.util.annotations.Thunk;

public class SwipeToReceiveFragment extends Fragment implements SwipeToReceiveViewModel.DataListener {

    private FragmentSwipeToReceiveBinding binding;
    @Thunk SwipeToReceiveViewModel viewModel;

    private BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(final Context context, final Intent intent) {
            if (intent.getAction().equals(BalanceFragment.ACTION_INTENT)) {
                if (viewModel != null) {
                    // Update UI with new Address + QR
                    viewModel.onViewReady();
                }
            }
        }
    };

    public SwipeToReceiveFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        // Listen for updates to stored addresses
        startWebSocketService();

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_swipe_to_receive, container, false);

        viewModel = new SwipeToReceiveViewModel(this);

        binding.qrCode.setOnClickListener(view -> showClipboardWarning());

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel.onViewReady();
    }

    @Override
    public void displayLoading() {
        binding.progressbar.setVisibility(View.VISIBLE);
        binding.qrCode.setVisibility(View.INVISIBLE);
    }

    @Override
    public void displayQrCode(Bitmap bitmap) {
        binding.progressbar.setVisibility(View.GONE);
        binding.qrCode.setVisibility(View.VISIBLE);
        binding.qrCode.setImageBitmap(bitmap);
    }

    @Override
    public void displayReceiveAddress(String address) {
        binding.receivingAddress.setText(address);

        // Register address as the one we're interested in via broadcast
        Intent intent = new Intent(WebSocketService.ACTION_INTENT);
        intent.putExtra("address", address);
        LocalBroadcastManager.getInstance(getContext()).sendBroadcast(intent);

        // Listen for corresponding broadcasts
        LocalBroadcastManager.getInstance(getContext()).registerReceiver(
                broadcastReceiver, new IntentFilter(BalanceFragment.ACTION_INTENT));
    }

    @Override
    public void displayReceiveAccount(String accountName) {
        binding.receivingAccount.setText(accountName);
    }

    @Override
    public void showNoAddressesAvailable() {
        binding.contentLayout.setVisibility(View.GONE);
        binding.errorLayout.setVisibility(View.VISIBLE);
    }

    private void showClipboardWarning() {
        new AlertDialog.Builder(getActivity(), R.style.AlertDialogStyle)
                .setTitle(R.string.app_name)
                .setMessage(R.string.receive_address_to_clipboard)
                .setCancelable(false)
                .setPositiveButton(R.string.yes, (dialog, whichButton) -> {
                    ClipboardManager clipboard = (android.content.ClipboardManager) getActivity().getSystemService(Context.CLIPBOARD_SERVICE);
                    ClipData clip = android.content.ClipData.newPlainText("Send address", binding.receivingAddress.getText().toString());
                    ToastCustom.makeText(getActivity(), getString(R.string.copied_to_clipboard), ToastCustom.LENGTH_LONG, ToastCustom.TYPE_GENERAL);
                    clipboard.setPrimaryClip(clip);
                })
                .setNegativeButton(R.string.no, null)
                .show();
    }

    @Override
    public void onStop() {
        super.onStop();
        LocalBroadcastManager.getInstance(getContext()).unregisterReceiver(broadcastReceiver);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        viewModel.destroy();
    }

    private void startWebSocketService() {
        Intent intent = new Intent(getContext(), WebSocketService.class);

        if (!new OSUtil(getActivity()).isServiceRunning(WebSocketService.class)) {
            getActivity().startService(intent);
        } else {
            // Restarting this here ensures re-subscription after app restart - the service may remain
            // running, but the subscription to the WebSocket won't be restarted unless onCreate called
            getActivity().stopService(intent);
            getActivity().startService(intent);
        }
    }
}

File: app/src/main/java/piuk/blockchain/android/ui/swipetoreceive/SwipeToReceiveFragment.kt
package piuk.blockchain.android.ui.swipetoreceive

import android.content.*
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v4.content.LocalBroadcastManager
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AlertDialog
import android.support.v7.widget.AppCompatImageView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import kotlinx.android.synthetic.main.fragment_swipe_to_receive.*
import piuk.blockchain.android.R
import piuk.blockchain.android.data.currency.CryptoCurrencies
import piuk.blockchain.android.data.websocket.WebSocketService
import piuk.blockchain.android.injection.Injector
import piuk.blockchain.android.ui.balance.BalanceFragment
import piuk.blockchain.android.ui.base.BaseFragment
import piuk.blockchain.android.ui.base.UiState
import piuk.blockchain.android.util.extensions.*
import piuk.blockchain.android.util.helperfunctions.setOnPageChangeListener
import javax.inject.Inject

@Suppress("MemberVisibilityCanPrivate")
class SwipeToReceiveFragment : BaseFragment<SwipeToReceiveView, SwipeToReceivePresenter>(),
    SwipeToReceiveView {

    @Inject lateinit var swipeToReceivePresenter: SwipeToReceivePresenter

    init {
        Injector.getInstance().presenterComponent.inject(this)
    }

    private val broadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == BalanceFragment.ACTION_INTENT) {
                // Update UI with new Address + QR
                presenter?.currencyPosition = presenter?.currencyPosition ?: 0
            }
        }
    }

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ) = container?.inflate(R.layout.fragment_swipe_to_receive)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        listOf(imageview_qr, textview_address, textview_request_currency).forEach {
            it.setOnClickListener { showClipboardWarning() }
        }

        imageview_left_arrow.setOnClickListener {
            viewpager_icons.currentItem = viewpager_icons.currentItem - 1
        }
        imageview_right_arrow.setOnClickListener {
            viewpager_icons.currentItem = viewpager_icons.currentItem + 1
        }

        val adapter = ImageAdapter(
                context!!,
                listOf(
                        R.drawable.vector_bitcoin,
                        R.drawable.vector_eth,
                        R.drawable.vector_bitcoin_cash
                )
        )

        viewpager_icons.run {
            offscreenPageLimit = 3
            setAdapter(adapter)
            indicator.setViewPager(this)
            setOnPageChangeListener {
                onPageSelected {
                    presenter.currencyPosition = it

                    when (it) {
                        0 -> imageview_left_arrow.invisible()
                        1 -> listOf(
                                imageview_left_arrow,
                                imageview_right_arrow
                        ).forEach { it.visible() }
                        2 -> imageview_right_arrow.invisible()
                    }
                }
            }
        }

        onViewReady()
    }

    override fun displayReceiveAddress(cryptoCurrency: CryptoCurrencies, address: String) {
        textview_address.text = address

        // Register address as the one we're interested in via broadcast
        val intent = Intent(WebSocketService.ACTION_INTENT).apply {
            when (cryptoCurrency) {
                CryptoCurrencies.BTC -> putExtra(WebSocketService.EXTRA_BITCOIN_ADDRESS, address)
                CryptoCurrencies.BCH -> putExtra(WebSocketService.EXTRA_BITCOIN_CASH_ADDRESS, address)
                CryptoCurrencies.ETHER -> Unit
            }
        }
        // No need to post ETH events as it should already have happened elsewhere
        if (cryptoCurrency != CryptoCurrencies.ETHER) {
            LocalBroadcastManager.getInstance(context!!).sendBroadcast(intent)
        }

        // Listen for corresponding broadcasts
        LocalBroadcastManager.getInstance(context!!).registerReceiver(
                broadcastReceiver, IntentFilter(BalanceFragment.ACTION_INTENT)
        )
    }

    override fun displayReceiveAccount(accountName: String) {
        textview_account?.text = accountName
    }

    override fun displayCoinType(requestString: String) {
        textview_request_currency?.text = requestString
    }

    override fun setUiState(uiState: Int) {
        when (uiState) {
            UiState.LOADING -> displayLoading()
            UiState.CONTENT -> showContent()
            UiState.FAILURE -> showNoAddressesAvailable()
            UiState.EMPTY -> showNoAddressesAvailable()
        }
    }

    override fun displayQrCode(bitmap: Bitmap) {
        imageview_qr.setImageBitmap(bitmap)
    }

    override fun onStop() {
        super.onStop()
        LocalBroadcastManager.getInstance(context!!).unregisterReceiver(broadcastReceiver)
    }

    override fun createPresenter() = swipeToReceivePresenter

    override fun getMvpView() = this

    private fun showContent() {
        layout_qr.visible()
        progress_bar.gone()
        imageview_qr.visible()
        textview_error.gone()
    }

    private fun displayLoading() {
        layout_qr.visible()
        progress_bar.visible()
        imageview_qr.invisible()
        textview_error.gone()
    }

    private fun showNoAddressesAvailable() {
        layout_qr.invisible()
        textview_error.visible()
        textview_address.text = ""
    }

    private fun showClipboardWarning() {
        activity?.run {
            AlertDialog.Builder(this, R.style.AlertDialogStyle)
                    .setTitle(R.string.app_name)
                    .setMessage(R.string.receive_address_to_clipboard)
                    .setCancelable(false)
                    .setPositiveButton(R.string.yes, { _, _ ->
                        val clipboard =
                                getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                        val clip = ClipData.newPlainText("Send address", textview_address.text)
                        toast(R.string.copied_to_clipboard)
                        clipboard.primaryClip = clip
                    })
                    .setNegativeButton(R.string.no, null)
                    .show()
        }
    }

    companion object {

        @JvmStatic
        fun newInstance(): SwipeToReceiveFragment = SwipeToReceiveFragment()

    }

    private class ImageAdapter(var context: Context, var drawables: List<Int>) : PagerAdapter() {

        override fun getCount(): Int = drawables.size

        override fun isViewFromObject(view: View, any: Any): Boolean {
            return view === any as AppCompatImageView
        }

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val imageView = container.inflate(R.layout.item_image_pager) as ImageView
            imageView.setImageDrawable(ContextCompat.getDrawable(context, drawables[position]))
            (container as ViewPager).addView(imageView, 0)
            return imageView
        }

        override fun destroyItem(container: ViewGroup, position: Int, any: Any) {
            container.removeView(any as LinearLayout)
        }

    }

}
